meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/ZkMetaClient.java

```diff
@@ -394,4 +414,69 @@ public byte[] serialize(T data, String path) {
   public T deserialize(byte[] bytes, String path) {
     return _zkClient.deserialize(bytes, path);
   }
+
+  /**
+   * A clean up method called when connect state change or MetaClient is closing.
+   * @param cancel If we want to cancel the reconnect monitor thread.
+   * @param close If we want to close ZkClient.
+   */
+  private void cleanUpAndClose(boolean cancel, boolean close) {
+    _zkClientConnectionMutex.lock();
+
+    if (close && !_zkClient.isClosed()) {
+      _zkClient.close();
+      LOG.info("ZkClient is closed");
+    }
+
+    if (cancel && _reconnectMonitorFuture != null) {
+      _reconnectMonitorFuture.cancel(true);
+      LOG.info("ZkClient reconnect monitor thread is canceled");
+    }
+
+    _zkClientConnectionMutex.unlock();
+  }
+
+  // Schedule a monitor to track ZkClient auto reconnect when Disconnected
+  // Cancel the monitor thread when connected.
+  @Override
+  public void handleStateChanged(Watcher.Event.KeeperState state) throws Exception {
+    if (state == Watcher.Event.KeeperState.Disconnected) {
+      // Expired. start a new event monitoring retry
+      _zkClientConnectionMutex.lockInterruptibly();
+      if (_reconnectMonitorFuture == null || _reconnectMonitorFuture.isCancelled()
+          || _reconnectMonitorFuture.isDone()) {
+        _reconnectMonitorFuture = _zkClientReconnectMonitor.schedule(() -> {
+          if (!_zkClient.getConnection().getZookeeperState().isConnected()) {
+            cleanUpAndClose(false, true);
+          }
+        }, _reconnectTimeout, TimeUnit.MILLISECONDS);
+        LOG.info("ZkClient is Disconnected, schedule a reconnect monitor after {}",
+            _reconnectTimeout);
+      }
+      _zkClientConnectionMutex.unlock();
+    } else if (state == Watcher.Event.KeeperState.SyncConnected
+        || state == Watcher.Event.KeeperState.ConnectedReadOnly) {
+      cleanUpAndClose(true, false);
+      LOG.info("ZkClient is SyncConnected, reconnect monitor thread is canceled (if any)");
+    }
```

Comment:
I have usually seen this at other places as well but just wondering should we add any error logs or some logging for "else" block here?

Reply:
ZkClient has log for other state change event. We are not doing anything related to reconnect monitor in other state.